Safer KTI termination

lutovich opened the pull request:
Previously it was possible for an external observer to get list of active transactions and terminate wrong ones. This could happen because of KTI pooling which means instance of KTI can represent multiple logical transactions during it's lifetime.

This commit makes active transactions listing return special objects and not KTI instances. These objects are not reused/pooled and guarantee that the right transaction will be terminated.

**Note:** I'll forward-merge this PR myself because `KernelTrasnactions#activeTransactions()` is used in more places in 3.1.

lutovich committed:
Safer KTI termination

Previously it was possible for an external observer to get list of active
transactions and terminate wrong ones. This could happen because of KTI
pooling which means instance of KTI can represent multiple logical transactions
during it's lifetime.

This commit makes active transactions listing return special objects and not
KTI instances. These objects are not reused/pooled and guarantee that the right
transaction will be terminated.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -242,6 +242,22 @@ public Status getReasonIfTerminated()
+        {
+            if ( expectedReuseCount == reuseCount )
+            {
+                markForTermination( reason );
```
common part from `markForTermination` should be extracted and invoked here instead to avoid double locking

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementationHandle.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public long lastTransactionTimestampWhenStarted()
+    {
+        return tx.lastTransactionTimestampWhenStarted();
```
This will leak timestamp from potentially not opened/ reused transaction. It breaks single logical transaction principle.

MishaDemianenko replied:
Also it rely on simple long field, so visibility of that data is questionable here as well

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -199,18 +200,25 @@ protected void dispose( KernelTransactionImplementation tx )
+    /**
+     * Create new handle for the given transaction.
+     * <p>
+     * <b>Note:</b> this method is package-private for testing <b>only</b>.
```
What about introducing @VisibleForTesting annotation instead?

lutovich replied:
I guess we can add such annotation but probably not in the maint release.

MishaDemianenko replied:
Do not see a problem with new test stuff in 3.0; but deal lets add it in 3.1

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementationHandle.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public boolean isOpen()
+    {
+        return tx.isOpen() && txReuseCount == tx.getReuseCount();
```
`tx.isOpen()` rely on simple booleans to keep it state, that will not be visible from other threads

lutovich replied:
Should be fixed now.

MishaDemianenko commented:
@lutovich first review round completed

lutovich committed:
Fix synchronization around KTI initialization, commit and termination

lutovich committed:
Memorize last commit timestamp in the tx handle

To make sure sane value is returned even if the underlying transaction has
been closed and possibly reused.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -204,19 +205,28 @@ public KernelTransactionImplementation( StatementOperationParts operations,
-        this.currentStatement.initialize( locks );
-        return this;
+        // guarded by the lock to coordinate with concurrent termination attempts
+        terminationReuseLock.lock();
```
locking here locks overkill, agreed to make locks volatile instead

lutovich committed:
Remove synchronization abound KTI#initialize()

Instead reference to it's Locks.Client is made volatile to be visible in
threads that try to terminate the KTI.

MishaDemianenko commented:
@lutovich review complete, feel free to merge it in when green.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -144,14 +144,14 @@ TransactionWriteState upgradeToSchemaWrites() throws InvalidTransactionTypeKerne
     private volatile Status terminationReason;
     private long startTimeMillis;
     private long lastTransactionIdWhenStarted;
-    private long lastTransactionTimestampWhenStarted;
```
That's a couple more fields made `volatile`, could we piggy-back some access on some other? Think about it, if we can't, we can't

MishaDemianenko replied:
we where also taking about some perf measurement so see influence of those

lutovich replied:
Here are some benchmark results: https://gist.github.com/lutovich/3077739e8628b2fd227c6f39a2ae1845

lutovich committed:
Memorize last commit id and start time in tx handle

To make sure sane values are returned even if the underlying transaction has
been closed and possibly reused.
